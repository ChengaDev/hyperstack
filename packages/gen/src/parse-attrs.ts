interface Attr {
  name: string
  as: As
}

interface As {
  fixtureAlt: string[]
  fixture: string[]
  model: string[]
  migration: string[]
  zod: string[]
}

type Conversion = (name: string) => string[]
interface TypeConversions {
  string: Conversion
  text: Conversion
  number: Conversion
  date: Conversion
  bool: Conversion
  json: Conversion
}

type Mappings = Record<keyof As, TypeConversions>

// XXX validate that 'name' is a valid symbol
// XXX validate that 'value' is a valid type, or that it has a translator
// XXX add a default | todo type for each area

const lines = (...lns: string[]) => lns

const mappings: Mappings = {
  // faker: {},
  fixtureAlt: {
    string: (_name) => ["'updated'"],
    text: (_name) => ["'a different text'"],
    number: (_name) => ['1337'],
    date: (_name) => ["new Date('1981-09-08')"],
    bool: (_name) => ['false'],
    json: (_name) => ['{"hello": "crewl world"}'],
  },
  fixture: {
    string: (_name) => ["'string'"],
    text: (_name) => ["'some text'"],
    number: (_name) => ['5150'],
    date: (_name) => ["new Date('1986-03-24')"],
    bool: (_name) => ['true'],
    json: (_name) => ['{"hello": "world"}'],
  },
  model: {
    string: (name) => lines('@Column', `${name}: string`),
    text: (name) => lines('@Column(DataType.TEXT)', `${name}: string`),
    number: (name) => lines('@Column', `${name}: number`),
    date: (name) => lines('@Column', `${name}: Date`),
    bool: (name) => lines('@Column', `${name}: boolean`),
    json: (name) => lines('@Column(DataType.JSON)', `${name}: any`),
  },
  migration: {
    string: (name) => lines(`t.string('${name}')`),
    text: (name) => lines(`t.text('${name}')`),
    number: (name) => lines(`t.int('${name}')`),
    date: (name) => lines(`t.date('${name}')`),
    bool: (name) => lines(`t.bool('${name}')`),
    json: (name) => lines(`t.json('${name}')`),
  },
  zod: {
    string: (name) => lines(`${name}: z.string()`),
    text: (name) => lines(`${name}: z.string()`),
    number: (name) => lines(`${name}: z.number()`),
    date: (name) => lines(`${name}: z.date()`),
    bool: (name) => lines(`${name}: z.boolean()`),
    json: (name) => lines(`${name}: z.any()`),
  },
}

const getMapping = (el: string, value: string) => {
  const f = mappings[el as keyof Mappings]?.[value as keyof TypeConversions]
  if (!f) {
    throw new Error(
      `type '${value}' not found. trying to convert '${value}' as '${el}'`
    )
  }
  return f
}
const parseAttr = (attr: string): Attr => {
  const props = attr.split(':')
  if (props.length !== 2) {
    throw new Error(
      `attribute: '${attr}' should be in the format [name]:[type]`
    )
  }
  const [name, value] = props

  // a somewhat conservative variable name. I know there's unicode and wild stuff _officially_
  // allowed, but let's be reasonable. This name will make its way into DB tables and other places.
  if (!value.match(/^[a-z]{1,}$/)) {
    throw new Error(`'${value}' is not a valid type name`)
  }
  if (!name.match(/^[a-zA-Z0-9_$]+$/)) {
    throw new Error(`'${name}' is not a valid variable name`)
  }

  return {
    name,
    as: Object.keys(mappings).reduce(
      (acc, el) => ({
        ...acc,
        [el]: getMapping(el, value)(name),
      }),
      {}
    ) as As,
  }
}
export const parseAttrs = (attrs: string | string[]): Attr[] => {
  if (!attrs) {
    throw new Error('no attributes to parse')
  }

  const parts = (attrs as string).split
    ? (attrs as string).split(/\s+/)
    : (attrs as string[])
  // 6 because that's the smallest type name (bool) + ':' + 1 char for the symbol name:
  // a:bool
  return parts.filter((p) => p.length >= 6 && p.includes(':')).map(parseAttr)
}
