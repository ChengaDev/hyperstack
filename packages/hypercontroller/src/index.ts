export * from './http/responses' // as Responses
export * from './validation'
export * from './errors'
export * from './types'
export * from './api'
export * from './jwt'
export * from './routes'
