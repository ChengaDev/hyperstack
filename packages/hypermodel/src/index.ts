import * as st from 'sequelize-typescript'
import { UUIDPK } from './hypermodel'

export const Schema = { ...st, UUIDPK } as typeof st & { UUIDPK: typeof UUIDPK }

export { HyperModel, HyperModelError, ModelAssertionError } from './hypermodel'
export * from './utils'
export { buildModels } from './models'
export * from './migrate'
export * from './converge'
export * from './types'
