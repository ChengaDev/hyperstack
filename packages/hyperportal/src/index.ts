/* eslint-disable no-console */
/* eslint-disable @typescript-eslint/no-var-requires */
import type { Context } from '@hyperstackjs/typings'
import asTable from 'as-table'
import L from 'lodash'
import { bold, gray, green, white } from 'colorette'
const repl = require('node:repl')

const table = asTable.configure({
  title: (x: any) => bold(green(x)),
  delimiter: gray(' | '),
  dash: gray('-'),
})

const serialize = (c: any, serializer: any) => {
  if (L.isArray(c)) {
    return L.map(c, serializer)
  } else {
    return serializer(c)
  }
}
const to = (c: any) => console.log(table(serialize(c, (v: any) => v.toJSON())))
const tr = (c: any) => console.log(table(serialize(c, (v: any) => v.get())))
const oo = (c: any) => console.log(serialize(c, (v: any) => v.toJSON()))
const ro = (c: any) => console.log(serialize(c, (v: any) => v.get()))

export const startPortal = (context: Context) => {
  const models = context.models()
  const controllers = context.controllers()
  const environment = context.environment()
  const printRoutes = context.helpers.printRoutes
  const printJobCounts = context.helpers.printJobCounts
  const replServer = repl.start({
    prompt: `âš¡[${bold(green(environment))}] ${bold(white('>'))} `,
  })
  Object.assign(replServer.context, models)
  Object.assign(replServer.context, { to, tr, oo, ro })

  replServer.defineCommand('jobs', {
    help: 'show job counts',
    async action() {
      this.clearBufferedCommand()
      await printJobCounts(context.workers())
      this.displayPrompt()
    },
  })
  replServer.defineCommand('models', {
    help: 'show models',
    action() {
      this.clearBufferedCommand()
      console.log(`${Object.keys(models).join('\n')}`)
      this.displayPrompt()
    },
  })
  replServer.defineCommand('routes', {
    help: 'show routes',
    action(snippets: string) {
      this.clearBufferedCommand()
      printRoutes(controllers, snippets ? [snippets] : undefined)
      this.displayPrompt()
    },
  })
  replServer.defineCommand('saybye', function saybye() {
    console.log('Goodbye!')

    // @ts-expect-error 2683
    this.close()
  })
}
