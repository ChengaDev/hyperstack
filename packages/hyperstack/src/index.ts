// expose the _right_ surface-area for users

export * from './tie'
export { initializer } from './initializers'
export { seed } from './models'
export { task } from './tasks'
export { isDevelopment, isProduction, context } from './context'
export { cli } from './cli'
export type {
  Context,
  Initializer,
  MigrationsMode,
  WorkersMode,
  Task,
  App,
  Config,
} from './types'

export {
  Schema,
  HyperModel,
  HyperModelError,
  ModelAssertionError,
  hashPassword,
  verifyPassword,
  generateHexToken,
} from '@hyperstackjs/hypermodel'

export {
  ClassErrorMiddleware,
  ClassMiddleware,
  Controller,
  Delete,
  ErrorMiddleware,
  Get,
  HttpResponse,
  HttpResponseBadRequest,
  HttpResponseClientError,
  HttpResponseConflict,
  HttpResponseCreated,
  HttpResponseForbidden,
  HttpResponseInternalServerError,
  HttpResponseMethodNotAllowed,
  HttpResponseMovedPermanently,
  HttpResponseNoContent,
  HttpResponseNotFound,
  HttpResponseNotImplemented,
  HttpResponseOK,
  HttpResponseRedirect,
  HttpResponseRedirection,
  HttpResponseServerError,
  HttpResponseSuccess,
  HttpResponseTooManyRequests,
  HttpResponseUnauthorized,
  Middleware,
  ParseBody,
  Patch,
  Post,
  Put,
  ValidateBody,
  ValidationError,
  authWithJWT,
  err,
  notfound,
  ok,
  permits,
  requires,
  unauthorized,
  Request,
  Response,
  NextFunction,
} from '@hyperstackjs/hypercontroller'

export {
  HyperWorker,
  Mailer,
  queueAs,
  producerOnly,
} from '@hyperstackjs/hyperworker'
