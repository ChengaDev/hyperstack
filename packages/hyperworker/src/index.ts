import { HyperWorker, producerOnly, queueAs } from './worker'
import { Mailer } from './mailer'
import { buildWorkers } from './hyperworker'
export * from './types'
export * from './ops'

export { HyperWorker, queueAs, Mailer, buildWorkers, producerOnly }
