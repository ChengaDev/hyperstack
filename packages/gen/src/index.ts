import path from 'path'
import { Logger, runner } from 'hygen'
import execa from 'execa'
import { parseAttrs } from './parse-attrs'
const defaultTemplates = path.join(__dirname, 'templates')

export const gen = (argv: string[]) => {
  runner(argv, {
    templates: defaultTemplates,
    cwd: process.cwd(),
    logger: new Logger(console.log.bind(console)), // eslint-disable-line no-console
    debug: !!process.env.DEBUG,
    exec: async (action, _body) => {
      // const opts = body && body.length > 0 ? { input: body } : {}
      const res = await execa(action, {
        shell: true,
      })
      return res
    },
    createPrompter: () => require('enquirer'),
    helpers: {
      parseAttrs,
    },
  }).then(({ success }) => process.exit(success ? 0 : 1))
}
