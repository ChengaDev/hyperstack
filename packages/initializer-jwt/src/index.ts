import { authWithJWT, err } from '@hyperstackjs/hypercontroller'
import type { Context } from '@hyperstackjs/typings'
import { context } from 'hyperstack'
import createDebug from 'debug'

import type { Request } from '@hyperstackjs/hypercontroller'
const debug = createDebug('hyperstack:init-jwt')

export const DEFAULT_AUTH_COOKIE_NAME = 'HS_AUTH'

export interface JWTProps {
  MustAuthWithJWT: any
  MustAuthRouteWithJWT: any
  signJWT: any
  verifyJWT: any
  currentUser: (req: any) => any
}
const key = 'jwt'
export const getProps = () => context.initializerProps<JWTProps>(key)

export const resolveAuthCookieName = (
  cookieNameConfig?: string | boolean
): string | null => {
  if (!cookieNameConfig) {
    return null
  }

  if (cookieNameConfig === true) {
    return DEFAULT_AUTH_COOKIE_NAME
  } else if (
    typeof cookieNameConfig == 'string' &&
    cookieNameConfig.length > 0
  ) {
    return cookieNameConfig
  } else {
    return null
  }
}

export default (
    createLoader: (context: Context) => (payload: any) => Promise<any | null>
  ) =>
  async (context: Context) => {
    const loader = createLoader(context)
    const config = context.config()
    const jwtSecretOrPublicKey = config.controllers!.jwtSecret
    if (!jwtSecretOrPublicKey) {
      throw new Error('No JWT secret is set in configuration')
    }
    const bearer = config.controllers!.bearer || undefined
    const resolvedAuthCookieName = resolveAuthCookieName(
      config.controllers!.authCookieName
    )

    let finalParams
    const baseParams = {
      secret: jwtSecretOrPublicKey,
      loader,
      bearer,
      jwtOptions: {
        expiresIn: config.controllers!.jwtExpiry || '14d',
        algorithm: config.controllers!.jwtAlgorithm || 'HS512',
      },
    }

    if (resolvedAuthCookieName) {
      finalParams = {
        ...baseParams,
        authCookieName: resolvedAuthCookieName,
      }
    } else {
      finalParams = {
        ...baseParams,
      }
    }

    const { MustAuthWithJWT, MustAuthRouteWithJWT, signJWT, verifyJWT } =
      authWithJWT(finalParams)

    return {
      provideProps() {
        return {
          [key]: {
            MustAuthWithJWT,
            MustAuthRouteWithJWT,
            signJWT,
            verifyJWT,
            currentUser(req: Request) {
              if (!req.user) {
                debug('no user in request')
                throw err('no user')
              }
              return req.user
            },
          },
        }
      },
    }
  }
