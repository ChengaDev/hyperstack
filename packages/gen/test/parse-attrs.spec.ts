import { parseAttrs } from '../src/parse-attrs'

describe('parseAttrs', () => {
  it('parses', () => {
    expect(
      parseAttrs(
        'g model Foobar title:string content:text count:number settings:json createdAt:date isDeleted:bool'
      )
    ).toMatchSnapshot()
    expect(
      parseAttrs(['g', 'model', 'title:string', 'content:text'])
    ).toMatchSnapshot()
    expect(() => parseAttrs(undefined as any)).toThrowError(/no attributes/)
    expect(() => parseAttrs('title:')).toThrowError(/not a valid/)
    expect(() => parseAttrs('title:footype')).toThrowError(
      /type 'footype' not found/
    )
    expect(() => parseAttrs('title:foo:bar')).toThrowError(/format/)
    expect(() => parseAttrs('ti-tle:string')).toThrowError(/not a valid/)
  })
  it('model but no attributes', () => {
    expect(parseAttrs('g model foobar')).toMatchSnapshot()
    expect(parseAttrs('model foobar')).toMatchSnapshot()
    expect(parseAttrs('foobar')).toMatchSnapshot()
  })
  it('typo testing', () => {
    expect(() => parseAttrs('title :string')).toThrowError(/not a valid/)
    expect(parseAttrs('title;string')).toEqual([])
    expect(() => parseAttrs('title: string')).toThrowError(/not a valid/)
    expect(() => parseAttrs('title:stringcontent :string')).toThrowError(
      /not found/
    )
    expect(() => parseAttrs('deleted:boolean')).toThrowError(/not found/)
    expect(() => parseAttrs('title.:string')).toThrowError(/not a valid/)
    expect(() => parseAttrs('title::string')).toThrowError(/format/)
  })
})
