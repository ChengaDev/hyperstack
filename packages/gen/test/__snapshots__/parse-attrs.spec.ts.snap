// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parseAttrs model but no attributes 1`] = `[]`;

exports[`parseAttrs model but no attributes 2`] = `[]`;

exports[`parseAttrs model but no attributes 3`] = `[]`;

exports[`parseAttrs parses 1`] = `
[
  {
    "as": {
      "fixture": [
        "'string'",
      ],
      "fixtureAlt": [
        "'updated'",
      ],
      "migration": [
        "t.string('title')",
      ],
      "model": [
        "@Column",
        "title: string",
      ],
      "zod": [
        "title: z.string()",
      ],
    },
    "name": "title",
  },
  {
    "as": {
      "fixture": [
        "'some text'",
      ],
      "fixtureAlt": [
        "'a different text'",
      ],
      "migration": [
        "t.text('content')",
      ],
      "model": [
        "@Column(DataType.TEXT)",
        "content: string",
      ],
      "zod": [
        "content: z.string()",
      ],
    },
    "name": "content",
  },
  {
    "as": {
      "fixture": [
        "5150",
      ],
      "fixtureAlt": [
        "1337",
      ],
      "migration": [
        "t.int('count')",
      ],
      "model": [
        "@Column",
        "count: number",
      ],
      "zod": [
        "count: z.number()",
      ],
    },
    "name": "count",
  },
  {
    "as": {
      "fixture": [
        "{"hello": "world"}",
      ],
      "fixtureAlt": [
        "{"hello": "crewl world"}",
      ],
      "migration": [
        "t.json('settings')",
      ],
      "model": [
        "@Column(DataType.JSON)",
        "settings: any",
      ],
      "zod": [
        "settings: z.any()",
      ],
    },
    "name": "settings",
  },
  {
    "as": {
      "fixture": [
        "new Date('1986-03-24')",
      ],
      "fixtureAlt": [
        "new Date('1981-09-08')",
      ],
      "migration": [
        "t.date('createdAt')",
      ],
      "model": [
        "@Column",
        "createdAt: Date",
      ],
      "zod": [
        "createdAt: z.date()",
      ],
    },
    "name": "createdAt",
  },
  {
    "as": {
      "fixture": [
        "true",
      ],
      "fixtureAlt": [
        "false",
      ],
      "migration": [
        "t.bool('isDeleted')",
      ],
      "model": [
        "@Column",
        "isDeleted: boolean",
      ],
      "zod": [
        "isDeleted: z.boolean()",
      ],
    },
    "name": "isDeleted",
  },
]
`;

exports[`parseAttrs parses 2`] = `
[
  {
    "as": {
      "fixture": [
        "'string'",
      ],
      "fixtureAlt": [
        "'updated'",
      ],
      "migration": [
        "t.string('title')",
      ],
      "model": [
        "@Column",
        "title: string",
      ],
      "zod": [
        "title: z.string()",
      ],
    },
    "name": "title",
  },
  {
    "as": {
      "fixture": [
        "'some text'",
      ],
      "fixtureAlt": [
        "'a different text'",
      ],
      "migration": [
        "t.text('content')",
      ],
      "model": [
        "@Column(DataType.TEXT)",
        "content: string",
      ],
      "zod": [
        "content: z.string()",
      ],
    },
    "name": "content",
  },
]
`;
